lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.h

```diff
@@ -33,12 +33,46 @@ class LLDBMemoryReader : public swift::remote::MemoryReader {
 
   void popLocalBuffer(); 
 
+  /// Adds the module to the list of modules we're tracking using tagged addresses,
+  /// so we can read memory from the file cache whenever possible.
+  /// \return a pair of addresses indicating the start and end of this image in the
+  /// tagged address space. None on failure.
+  llvm::Optional<std::pair<uint64_t, uint64_t>> addModuleToAddressMap(lldb::ModuleSP module);
+
+private:
+  /// Resolves the address by either mapping a tagged address back to an LLDB 
+  /// Address with section + offset, or, in case the address is not tagged, 
+  /// constructing an LLDB address with just the offset.
+  /// \return an Address with Section + offset  if we succesfully converted a tagged
+  /// address back, an Address with just an offset if the address was not tagged,
+  /// and None if the address was tagged but we couldn't convert it back to an 
+  /// Address.
+  llvm::Optional<Address> resolveRemoteAddress(uint64_t address) const;
+
 private:
   Process &m_process;
   size_t m_max_read_amount;
 
   llvm::Optional<uint64_t> m_local_buffer;
   uint64_t m_local_buffer_size = 0;
+
+  // A map from tagged addresses that we hand out to modules that
+  // LLDBMemoryReader will read from.
+  // We map the start address of the next module (if it's being used, otherwise
+  // it's the next available address) to the current module. This is done so
+  // we can tell what the upper bound of all mapped images is.
+  // Addresses are always inserted in ascending order, so the data structure is
+  // guaranteed to be sorted.
+  std::vector<std::pair<uint64_t, lldb::ModuleSP>> m_range_module_map;
+
+  // Tagged addresses we hand out will contain this mask to indicate that
```

Comment:
/// The bit used to tag LLDB's virtual addresses as such. See \c m_lldb_virtual_address_map.